naming/pubsub.md

```diff
@@ -0,0 +1,94 @@
+# ![](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square) IPNS PubSub Router
+
+Authors:
+
+  - Adin Schmahmann ([@aschmahmann](https://github.com/aschmahmann))
+  
+Reviewers:
+
+-----
+  
+# Abstract
+
+[Inter-Planetary Naming System (IPNS)](/README.md) is a naming system responsible for the creating, reading and updating of mutable pointers to data.
+IPNS consists of a public/private asymmetric cryptographic key pair, a record type and a protocol.
+Part of the protocol involves a routing layer that is used for the distribution and discovery of new or updated IPNS records.
+
+The IPNS PubSub router uses [libp2p PubSub](https://github.com/libp2p/specs/tree/master/pubsub) as a base, and adds persistence on top of it to ensure IPNS updates are always available to a connected network.
+An inherent property of the IPNS PubSub Router is that IPNS records are republishable by peers other than the peer that originated the record.
+This implies that as long as a peer on the network has an IPNS record it can be made available to other peers (although the records may be ignored if they are received after the IPNS record's End-of-Life/EOL).
+
+# Organization of this document
+
+  - [Introduction](#introduction)
+  - [Protocol Overview](#pubsub-protocol-overview)
+  - [Protocol](#protocol)
+  - [Implementations](#implementations)
+
+# Introduction
+
+Each time a node publishes an updated IPNS record for a particular key it is propagated by the router into the network where network nodes can choose to accept or reject the new record.
+When a node attempts to retrieve an IPNS record from the network it uses the router to query for the IPNS record(s) associated with the IPNS key; the node then validates the received records.
+
+In this spec we address building a router based on a PubSub system, particularly focusing on libp2p PubSub.
+
+# PubSub Protocol Overview
+
+The protocol has four components:
+- [IPNS Records and Validation](/README.md)
+- [libp2p PubSub](https://github.com/libp2p/specs/tree/master/pubsub)
+- Translating an IPNS record name to/from a PubSub topic
+- Layering persistence onto libp2p PubSub
+
+# Translating an IPNS record name to/from a PubSub topic
+
+For a given IPNS local record key described in the IPNS Specification the PubSub topic is:
+
+**Topic format:** `/record/base64url-unpadded(key)`

```

Comment:
Good point about the inconsistency, there's two issues here:

1) I think that area of the IPNS spec could use some clarification. Here, we're actually referring to the [ipfs/specs/naming#routing-record](https://github.com/ipfs/specs/tree/a370a6aa2345b251a4c04d37b8c50d3aa8f2701f/naming#routing-record) `/ipns/BINARY_ID`.
2) This spec is currently using `/record` instead of `/ipns`. Should we make it use `/ipns`? If starting with `/ipns` is not necessary for routing records across multiple routers then we should probably just remove the section on routing records.

What do you think?

Reply:
I asked @Stebalien what he thought about this and his understanding of the linked IPNS is that:

local-record = what we put in a local datastore = `/ipns/base32(key)` where `key = multihash of IPNS public key`
routing-record = what the form of the routing record is (`BINARY_ID = binary representation of key` where `key = multihash of IPNS public key`

This means our topic format is `/record/base64url-unpadded(BINARY_ID)`. However, we should give BINARY_ID a new name in the other spec that makes more sense. One set of words we tend to use is `IPNS private key = signing key, IPNS public key = verification key, IPNS key = multihash(IPNS public key)`. However, we could probably use better words, any suggestions?